### DISCUSSION TITLE
Prisma, Postgres, QueryRaw and RETURNING

### DISCUSSION THREAD

**Discussion by k7-soeren** [prisma/prisma#23294] [Q&A] [answered]
### Question

Hi there,
I am using Prisma as my main-db-connector and am using the full potential with "my" tables. As I have to work on a db, that already has tables that I need to interact with, I need to use queryRaw and executeRaw because those tables are not under my schema-control.

Now if I perform an INSERT I am utilising the "RETURNING" syntax to get the last inserted id, but `queryRaw` always returns the rows touched/inserted even when using RETURNING statement.

How would I fix that (w/o having to select the last inserted id "manually")?

Keep up the good work!

Thanks

### How to reproduce (optional)

Table `mytable`:
| id | mycol |
|--|--|
| a sequence (nextval would be 18) | smallint |

```ts
const insertId = await prismaClient.$executeRaw<number>`INSERT INTO mytable (mycol) VALUES (250) RETURNING id`;
```

insertId is always 1 regardless of sequence counter

### Expected behavior (optional)

well, executeRaw does what it says in the docs, but I'd like to get the RETURNING id as well :)

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Win10
- Database: Postgres 14
- Node.js version: 18.15.0

`prisma -v`
prisma                  : 5.9.1
@prisma/client          : 5.9.1
Computed binaryTarget   : windows
Operating System        : win32
Architecture            : x64
Node.js                 : v18.15.0
Query Engine (Node-API) : libquery-engine 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 23fdc5965b1e05fc54e5f26ed3de66776b93de64 (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.9.0-32.23fdc5965b1e05fc54e5f26ed3de66776b93de64
Default Engines Hash    : 23fdc5965b1e05fc54e5f26ed3de66776b93de64
Studio                  : 0.497.0

**Comment by ludralph** [comment, marked as answer]
Hi @k7-soeren 

Thank you for raising this question.

The behavior you're experiencing, where `executeRaw` returns the number of affected rows instead of the actual rows or IDs from a RETURNING clause, is consistent with Prisma's design.

You might want to use `queryRaw` and then extract the ID from the returned array. 

```js
const result = await prismaClient.$queryRaw`INSERT INTO mytable (mycol) VALUES (250) RETURNING id`;
const insertId = result[0].id; // Assuming the ID is returned in the first object of the array
```

**Comment by k7-soeren** [comment]
Hi @ludralph 

yes, `executeRaw` works as expected. I would've never even considered swapping execute with query.

Are you able to answer if that affects a transaction in any way?

I have opened a transaction with the 
```ts
await prisma.$transaction(async (prismaClient) => {
```
syntax and just want to make sure changing `execute` with `query` won't have any side effects while inside the transaction.


Thanks

**Reply by ludralph** [reply]
Both `$executeRaw` and `$queryRaw` can be used within transactions and there is no side effects.
